presto-main/src/main/java/com/facebook/presto/operator/Driver.java

```diff
@@ -258,35 +268,55 @@ private void processNewSources()
     {
         checkLockNotHeld("Can not process while holding the driver lock");
 
+        // if the driver is blocked we don't need to continue
+        if (!previousDriverBlockedFuture.isDone()) {
+            return previousDriverBlockedFuture;
+        }
+
         Optional<ListenableFuture<?>> result = tryWithLock(100, TimeUnit.MILLISECONDS, () -> {
             ListenableFuture<?> driverBlockedFuture = processInternal();
-            return wakeUpOnRevokeRequest(driverBlockedFuture);
+            previousDriverBlockedFuture = wakeUpOnRevokeRequest(driverBlockedFuture);
+            return previousDriverBlockedFuture;
         });
         return result.orElse(NOT_BLOCKED);
     }
 
     private ListenableFuture<?> wakeUpOnRevokeRequest(ListenableFuture<?> driverBlockedFuture)
     {
+        verify(previousDriverBlockedFuture.isDone(), "previousDriverBlockedFuture must be done");
+
         if (driverBlockedFuture.isDone()) {
             // driver is not blocked; just return completed future
             return driverBlockedFuture;
         }
 
-        ImmutableList<SettableFuture<?>> revokingRequestedFutures = operators.stream()
+        ImmutableList<MemoryRevokingRequestedFuture<?>> revokingRequestedFutures = operators.stream()
                 .map(Operator::getOperatorContext)
                 .map(OperatorContext::getMemoryRevokingRequestedFuture)
                 .collect(toImmutableList());
-        Optional<SettableFuture<?>> doneRevokingRequestedFuture = revokingRequestedFutures.stream().filter(Future::isDone).findAny();
+
+        Optional<MemoryRevokingRequestedFuture<?>> doneRevokingRequestedFuture = revokingRequestedFutures.stream()
+                .filter(MemoryRevokingRequestedFuture::isDone)
+                .findAny();
+
         if (doneRevokingRequestedFuture.isPresent()) {
-            // revoking already requested for one of operators; return completed future
-            return doneRevokingRequestedFuture.get();
+            // revoking is already requested for one of the operators; return completed future
+            return Futures.immediateFuture(null);
         }
 
-        // unblock as soon as driver is unblocked or we get revoking request for any of operators
-        ImmutableList.Builder<ListenableFuture<?>> futures = ImmutableList.builder();
-        futures.add(driverBlockedFuture);
-        futures.addAll(revokingRequestedFutures);
-        return firstFinishedFuture(futures.build());
+        // we can clear the listeners now before registering the new ones below
+        revokingRequestedFutures.stream()
+                .filter(MemoryRevokingRequestedFuture::isListenerSet)
+                .forEach(MemoryRevokingRequestedFuture::unsetListener);
```

Comment:
The biggest problem i have with `MemoryRevokingRequestedFuture` class is that I don't understand it's purpose. Of course, this might be just me, but why aren't changes to Driver (~ `Driver#previousDriverBlockedFuture`) a sufficient fix? I am aware that if driver blocks repeatedly, one innocent Operator's memory revoking requested future will get more and more listeners, but maybe this "singletonness of a listener" could be done in Driver:

1. driver keeps track of a isDriverBlockedFuture last returned (= `previousDriverBlockedFuture` but with clear intent that this must be accurate, not just for saving memory)
2. driver has one listener instance that, upon being called completes last isDriverBlockedFuture (with proper synchronization)
3. driver registers this singleton listener:
** upon construction, in each operator's memory revoking requested future
** when  calling operator's reset memory revoking requested

Wouldn't that be sufficient?

Of course, if we go this way, this should be reinforced by API changes to make the Driver-Operator interaction clear: like making `resetMemoryRevokingRequested` accept this listener as an argument.

I know this is kind if coupling Driver-Operator, but current code is not free of such coupling either. Limiting to singleton listener means no-one else can listen to memory revoking requested as Driver already does this.

Reply:
>> why aren't changes to Driver (~ Driver#previousDriverBlockedFuture) a sufficient fix?

Even with that when no memory revoking happens the memory revoking requested futures will never complete and the listeners added to them in `wakeUpOnRevokeRequest` may never be removed. This check guarantees that it's safe to set the single listener on the new memory revoke requested feature (but we now introduce a limitation of being able to add only a single listener as you point below).

Reply:
I put it wrongly, sorry for that. I am ware just having this field is not sufficient. I've, however, outlined a proposal for an alternative approach, one that does not require introduction of a specialized class (the `MemoryRevokingRequestedFuture`).
What do you think about it?

Reply:
@nezihyigitbasi If you are fine with that I will try to code what @findepi proposed and see how it plays with the codebase.

Reply:
@nezihyigitbasi Here is sth inline with @findepi thinking (I believe). 

https://github.com/starburstdata/presto/commits/lo/driver-single-listener-revoking

It is lacking tests etc. So it is not ready for PR of course. But if you see the approach fit we can extend this.